### DISCUSSION TITLE
Accessing Lists Field...

### DISCUSSION THREAD

**Discussion by afk-Legacy** [vfsfitvnm/frida-il2cpp-bridge#36] [Q&A]
I intend to access a field which is a list, I am pretty new to ts and frida in genreal so not sure how to do so,

![image](https://user-images.githubusercontent.com/83068093/133591078-e8b3e7f2-cb83-4079-9171-cbd74a90bf76.png)

**Comment by vfsfitvnm** [comment]
You have to handle it like a general `Il2CppObject`.
`System.Collections.Generic.List<T>` is a `Il2CppObject` under any point of view - it doesn't have a corresponding `Il2CppTypeEnum` so it doesn't have to be handled specially.

```ts

const Point = object.fields.Point;

// System.Collections.Generic.List<System.Int64>
console.log(Point.type.name);

// it will print the System.Collections.Generic.List<System.Int64> class
console.log(Point.type.class);

const PointValue = Point.value as Il2Cpp.Object;

const size = PointValue.methods.get_Count.invoke<number>();
PointValue.methods.Add.invoke<void>(1234);
const newSize = PointValue.methods.get_Count.invoke<number>();
console.log(size + 1 == newSize);
```

**Comment by afk-Legacy** [comment]
```
pointInString.implementation = function (format : Il2Cpp.String) : Il2Cpp.String
    {
        
        const PointValue = this.fields.Point.value as Il2Cpp.Object;
        
        const val = PointValue.methods.get_Item.invoke<number>(0);


        //log(JSON.stringify(state));
        return this.methods["ToString_"].invoke<Il2Cpp.String>(format);
    }
```
Generating Error: 

```

Error: expected a NativePointer object
    at NativeFunction.<anonymous> (<anonymous>)
    at Il2CppMethod.invokeRaw (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:141:1)
    at Il2CppMethod.<anonymous> (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:149:1)
    at Il2CppObject.pointInString.implementation (index.ts:40:49)
    at InvocationContext.replaceCallback (../../node_modules/frida-il2cpp-bridge/dist/il2cpp/structs/method.js:95:1)

```

**Reply by vfsfitvnm** [reply]
I can't tell much from this.
If you are on Android, could you tell me the app name?

**Reply by afk-Legacy** [reply]
https://play.google.com/store/apps/details?id=com.example.application

**Reply by vfsfitvnm** [reply]
Ok, could you provide more context now?
I can't find `pointInString` method. Also, how it is triggered?

**Reply by afk-Legacy** [reply]
hey sorry for the late reply, the class is EgoPoint and the hooked function is `public string ToString(string format)` , there is only one field here whose data I am trying to log.

**Reply by afk-Legacy** [reply]
The ToString is always being called in game so you wont have much of difficulty managing to execute it.

**Reply by afk-Legacy** [reply]
Also, how can we get the address of a variable, for example I want to find the address of variable "Point" in memory, navigating to the provided memory address would reveal the contents of the list, size and so on....

**Reply by vfsfitvnm** [reply]
It looks like a bug that happens with inflated methods.
Calling an inflated method requires the corresponding `Il2CppMethod` to be appended to the parameter list, but this is not true for every inflated method (I don't know why, maybe it is only true for constructors?).

However, the bug occurred because I was prepending (instead of appending) that extra `"pointer"` type parameter:
for `get_Item`, I was generating `["pointer", "pointer", "int32"]` instead of `["pointer", "int32", "pointer"]` (instance, index, method). So that's why Frida got angry. You can check the generated signature yourself via `Il2Cpp.Method.fridaSignature`.

It is funny how this bug wasn't discovered earlier (probably a regression?).

I'll push a new version as soon as I can, I want to investigate a little more, first.

Meanwhile, you can declare your own `get_Item` like this:
```ts
const AssemblyCSharp = Il2Cpp.Domain.assemblies["Assembly-CSharp"].image;
const EgoPoint = AssemblyCSharp.classes.EgoPoint;

const get_Item = new NativeFunction(EgoPoint.fields.Point.type.class.methods.get_Item.virtualAddress, "int64", ["pointer", "int32"]); // in this case, that statement it's not true

EgoPoint.methods.ToString_.implementation = function (format: Il2Cpp.String): Il2Cpp.String {
    const Point = this.fields.Point.value as Il2Cpp.Object;
        
    const firstValue: Int64 = get_Item.invoke(Point, 0);

    return this.methods.ToString_.invoke(format);
};
```

**Reply by vfsfitvnm** [reply]
Fixed in `0.5.6`
